File: Framework/Data Manipulation/BaseQueryObject.cs
using System;
using System.Collections.Generic;

namespace Framework.Data_Manipulation
{
    public interface IBaseQueryObject
    {
        Type SearchInputType { get; }
        object SearchInputObject { get; set; }
    }

    public interface IBaseQueryObject<TEntity> : IBaseQueryObject
    {
        IList<TEntity> Execute();
    }

    public interface IBaseQueryObject<TEntity, TSearchInput> : IBaseQueryObject<TEntity>
    {
        TSearchInput SearchInput { get; set; }
        IList<TEntity> PerformSearchOperation(TSearchInput searchInput);
    }

    public abstract class BaseQueryObject<TEntity, TSearchInput> :
        IBaseQueryObject<TEntity, TSearchInput>
    {
        public TSearchInput SearchInput { get; set; }

        public object SearchInputObject
        {
            get { return SearchInput; }
            set { SearchInput = (TSearchInput) value; }
        }

        public Type SearchInputType {
            get { return typeof(TSearchInput); }
        }

        public abstract IList<TEntity> PerformSearchOperation(TSearchInput searchInput);

        public IList<TEntity> Execute()
        {
            IList<TEntity> ultimateResult = PerformSearchOperation(SearchInput);

            return ultimateResult;
        }
    }
}

File: Android/framework/src/main/java/com/DataManipulation/BaseQueryObject.java
package com.DataManipulation;

import com.DataManipulation.BaseQueryObjectInterfaces.IBaseQueryObjectSearchableConcrete;
import java.util.List;

/**
 * Created by aiko on 2/11/17.
 */

public abstract class BaseQueryObject<TEntity, TSearchInput>
        implements IBaseQueryObjectSearchableConcrete<TEntity, TSearchInput> {
    TSearchInput _searchInput;
    Class _searchInputType;

    public TSearchInput GetSearchInput() {
        return _searchInput;
    }

    public void SetSearchInput(TSearchInput searchInput) {
        _searchInput = searchInput;
    }

    public Object GetSearchInputObject() {
        return _searchInput;
    }

    public void SetSearchInputObject(Object inputObject) {
        _searchInput = (TSearchInput) inputObject;
    }

    public Class GetSearchInputType() {
        return _searchInputType;
    }

    public BaseQueryObject(Class searchInputType) {
        _searchInputType = searchInputType;
    }

    public abstract List<TEntity> PerformSearchOperation(TSearchInput searchInput);

    public List<TEntity> Execute() {
        List<TEntity> searchResult = PerformSearchOperation(_searchInput);

        return searchResult;
    }
}
